### DISCUSSION TITLE
[GUIDE] Sync local and prod schemas when they're out of sync

### DISCUSSION THREAD

**Discussion by techwizzdom** [supabase/supabase#18483] [Show and tell]
Hey all 👋🏼,

I noticed that many people have problems when they make changes in remote DB and local DB only by clicking around the Supabase UI - and then, projects go out of sync. As a consequence, you cannot make nice migrations with `db diff/push`.

I had the same problem, but I finally found a solution I don't see it in the documentation, so I'm sharing it here:

1. Let's say that your remote DB is in a desired state. You need to get rid of all the migrations you previously did. In your remote SQL editor, do:
```
delete from supabase_migrations.schema_migrations;
```

2. Now, fetch the latest remote DB schema with:
```
supabase db pull
```
This will create a remote schema in `supabase/migrations/{{timestamp}}_remote_schema.sql`.

3. It's time to see the difference between your local and remote:
```
supabase db diff -f initial_diff
```
This will create a difference between local and remote DB schemas in `supabase/migrations/{{timestamp}}_initial_diff.sql`.

4. Go through `*_initial_diff.sql` file and see the differences. Keep only those you want to apply to the remote.

5. When you're ready, do:
```
supabase db push --dry-run
```
This will show which migrations will be applied to the remote, **it will not actually push them. It's just for checking**. If you did everything right, you should see only the `initial_diff` migration:
```
✗ supabase db push --dry-run
DRY RUN: migrations will *not* be pushed to the database.
Would push migration {{timestamp}}_initial_diff.sql...`
Finished supabase db push.
```

6. Now, you're ready to push the migrations:
```
supabase db push
```
Check your remote - changes in initial_diff should be applied.

7. And from now on, the process is pretty simple. You do some changes in your local DB (through the UI or SQL editor, doesn't really matter). Let's say we added a column `new_column` to table `screenplays` via UI. If you then do:
```
supabase db diff -f add_new_column
```
A new file `supabase/migrations/{{timestamp}}_add_new_column.sql` will be created and inside you should see only:
```
alter table "public"."screenplays" add column "new_column" text;
```
And changes you kept from the initial diff (if any).

8. If that's the case, it's all synced and ready! 🚀

It would be awesome to add this to the official docs.

**Comment by Julian-Tipler** [comment]
Thank you so much for this explanation! Follow-up question, do we check the remote_schema.sql into version control?

**Reply by stamler** [reply]
Yes. It's the source of truth for your schema.

**Comment by monaye** [comment]
> Let's say that your remote DB is in a desired state. You need to get rid of all the migrations you previously did. In your remote SQL editor, do:  

so, basically, after you finished pushing, you can delete the generated file right away ?!

**Reply by charliemday** [reply]
Yeah I had a similar question. Do you (@techwizzdom) mean your local SQL editor because surely you don't want to delete the migrations in your remote DB?

**Comment by Rubenburdin** [comment]
This is a solid writeup and addresses something thats bitten a lot of teams. The confusion in the replies about whether to delete migrations locally or remotely is worth clarifying though, I think you meant the remote supabase_migrations.schema_migrations table specifically to reset the migration history there, not delete the actual migration files.
One thing to be careful with in 2025, deleting from that migrations table works but it essentially tells Supabase "forget what migrations have been applied." If you have a team or CI pipeline thats also tracking migrations, this can cause conflicts where someone else tries to push and gets unexpected results. For teams with multiple developers, its usually safer to treat production schema as the source of truth and regenerate your local migration history from it using db pull, rather than clearing the remote history.
Also worth noting that supabase db diff compares against your local database state, so if your local isnt running or has drifted in unexpected ways the diff output can be misleading. Running supabase db reset before diffing ensures youre comparing against a clean baseline built from your migration files.
To answer Julian and monaye's questions, yes the remote_schema.sql should go into version control as your baseline. And you dont delete the migration files after pushing, those stay in your repo as your schema history. The only thing being "cleared" in step 1 is the migration tracking table so Supabase doesnt think old migrations still need to run.
For data sync issues between environments rather than schema sync, thats a different problem entirely and tools like Stacksync approach it from the angle of keeping actual records consistent across systems. But for schema management what you've outlined here is the right workflow.
